onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h

```diff
@@ -37,6 +44,38 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
   }
 
   virtual std::shared_ptr<KernelRegistry> GetKernelRegistry() const override;
+
+  std::shared_ptr<mkldnn::memory> GetWeightsMemory(const std::string& weight_key) {
```

Comment:
can we make GetWeightsMemory(), SetWeightsMemory(), GetMutex(), and SaveAllocatedMemory() private
they shouldn't be exposed externally. 

and we can get rid of the lock_guard in GetWeightsMemory and SetWeightsMemory